File: find_dups_go/find_dups.go
// find_dups.go
package main

import (
	"crypto/sha256"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"sync"
	"sync/atomic"
	"syscall"
	"time"
)

type FileInfo struct {
	ID      int
	Path    string
	Size    int64
	ModTime time.Time
	Birth   time.Time
	Hash    string
}

var categoryMap = map[string]string{
	".c": "source", ".h": "source", ".cpp": "source", ".hpp": "source",
	".cc": "source", ".cxx": "source", ".m": "source", ".mm": "source",
	".s": "source", ".S": "source", ".java": "source", ".kt": "source",
	".py": "source", ".js": "source", ".ts": "source", ".rs": "source",
	".go": "source", ".rb": "source", ".swift": "source", ".sh": "source",

	".hex": "firmware", ".bin": "firmware", ".elf": "firmware", ".dfu": "firmware",
	".flash": "firmware", ".map": "firmware",

	".uvprojx": "ide", ".uvoptx": "ide", ".ewp": "ide", ".eww": "ide",
	".ewt": "ide", ".cproject": "ide", ".project": "ide", ".mxproject": "ide", ".ioc": "ide",

	".yaml": "config", ".yml": "config", ".cmake": "config", ".json": "config",
	".xml": "config", ".conf": "config", ".cfg": "config", ".ini": "config",
	".toml": "config", ".properties": "config",

	".pdf": "docs", ".md": "docs", ".txt": "docs", ".html": "docs",
	".htm": "docs", ".rst": "docs", ".chm": "docs", ".doc": "docs",
	".docx": "docs", ".rtf": "docs",

	".png": "image", ".jpg": "image", ".jpeg": "image", ".gif": "image",
	".webp": "image", ".svg": "image", ".bmp": "image", ".ico": "image",
	".tiff": "image", ".tif": "image",

	".exe": "binary", ".dll": "binary", ".so": "binary", ".dylib": "binary",
	".o": "binary", ".a": "binary", ".lib": "binary", ".obj": "binary",
	".gch": "binary", ".pch": "binary",

	".zip": "archive", ".7z": "archive", ".tar": "archive", ".gz": "archive",
	".bz2": "archive", ".xz": "archive", ".rar": "archive", ".tgz": "archive",

	".mp4": "media", ".wav": "media", ".avi": "media", ".mp3": "media",
	".ogg": "media", ".flac": "media", ".mov": "media", ".wmv": "media",

	".ttf": "font", ".otf": "font", ".woff": "font", ".woff2": "font",

	".csv": "data", ".dts": "data", ".dtsi": "data", ".overlay": "data",
	".ld": "data", ".icf": "data", ".srec": "data",
}

// ProgressTracker tracks and displays file-based progress every 1 second with spinner animation
type ProgressTracker struct {
	total      int64
	processed  int64
	itemName   string
	startTime  time.Time
	stopChan   chan struct{}
}

func NewProgressTracker(total int, itemName string) *ProgressTracker {
	return &ProgressTracker{
		total:     int64(total),
		processed: 0,
		itemName:  itemName,
		startTime: time.Now(),
		stopChan:  make(chan struct{}),
	}
}

func (pt *ProgressTracker) Start() {
	ticker := time.NewTicker(1 * time.Second)
	spinners := []string{"|", "/", "-", "\\"}
	spinIdx := 0
	go func() {
		for {
			select {
			case <-ticker.C:
				current := atomic.LoadInt64(&pt.processed)
				if current >= pt.total {
					return
				}
				spinChar := spinners[spinIdx]
				spinIdx = (spinIdx + 1) % len(spinners)
				fmt.Printf("\n\rHashing: %d/%d files %s", current, pt.total, spinChar)
			case <-pt.stopChan:
				return
			}
		}
	}()
}

func (pt *ProgressTracker) Increment() {
	atomic.AddInt64(&pt.processed, 1)
}

func (pt *ProgressTracker) Stop() {
    close(pt.stopChan)
    fmt.Printf("\rHashing: %d/%d files \u2588\n", pt.total, pt.total)
}

func getCategory(ext string) string {
	if cat, ok := categoryMap[ext]; ok {
		return cat
	}
	return "other"
}

func generateAnalytics(allFiles []FileInfo, bySize map[int64][]*FileInfo, elapsed time.Duration) {
	// Identify duplicate file IDs
	dupSet := make(map[int]bool)
	for _, files := range bySize {
		if len(files) < 2 {
			continue
		}
		hashGroups := make(map[string][]*FileInfo)
		for _, f := range files {
			if f.Hash != "" {
				hashGroups[f.Hash] = append(hashGroups[f.Hash], f)
			}
		}
		for _, group := range hashGroups {
			if len(group) < 2 {
				continue
			}
			sort.Slice(group, func(i, j int) bool { return group[i].ID < group[j].ID })
			for i := 1; i < len(group); i++ {
				dupSet[group[i].ID] = true
			}
		}
	}

	// Accumulate stats
	type catStats struct {
		count         int
		totalBytes    int64
		dupCount      int
		dupBytes      int64
		extensions    map[string]int
	}
	type extStats struct {
		count      int
		totalBytes int64
		dupCount   int
		dupBytes   int64
	}

	catMap := make(map[string]*catStats)
	extMap := make(map[string]*extStats)

	var totalSize int64
	var dupSize int64
	sizeDist := map[string]int{
		"0_bytes":   0,
		"under_1kb": 0,
		"1kb_100kb": 0,
		"100kb_1mb": 0,
		"1mb_100mb": 0,
		"over_100mb": 0,
	}

	for i := range allFiles {
		f := &allFiles[i]
		ext := strings.ToLower(filepath.Ext(f.Path))
		cat := getCategory(ext)
		isDup := dupSet[f.ID]

		totalSize += f.Size
		if isDup {
			dupSize += f.Size
		}

		// Category stats
		cs, ok := catMap[cat]
		if !ok {
			cs = &catStats{extensions: make(map[string]int)}
			catMap[cat] = cs
		}
		cs.count++
		cs.totalBytes += f.Size
		if isDup {
			cs.dupCount++
			cs.dupBytes += f.Size
		}
		if ext != "" {
			cs.extensions[ext]++
		}

		// Extension stats
		es, ok := extMap[ext]
		if !ok {
			es = &extStats{}
			extMap[ext] = es
		}
		es.count++
		es.totalBytes += f.Size
		if isDup {
			es.dupCount++
			es.dupBytes += f.Size
		}

		// Size distribution
		sz := f.Size
		switch {
		case sz == 0:
			sizeDist["0_bytes"]++
		case sz < 1024:
			sizeDist["under_1kb"]++
		case sz < 100*1024:
			sizeDist["1kb_100kb"]++
		case sz < 1024*1024:
			sizeDist["100kb_1mb"]++
		case sz < 100*1024*1024:
			sizeDist["1mb_100mb"]++
		default:
			sizeDist["over_100mb"]++
		}
	}

	// Build JSON structure
	byCategory := make(map[string]interface{})
	for cat, cs := range catMap {
		exts := make(map[string]int)
		for e, n := range cs.extensions {
			exts[e] = n
		}
		byCategory[cat] = map[string]interface{}{
			"count":          cs.count,
			"total_bytes":    cs.totalBytes,
			"duplicate_count": cs.dupCount,
			"duplicate_bytes": cs.dupBytes,
			"extensions":     exts,
		}
	}

	byExtension := make(map[string]interface{})
	for ext, es := range extMap {
		byExtension[ext] = map[string]interface{}{
			"count":           es.count,
			"total_bytes":     es.totalBytes,
			"duplicate_count": es.dupCount,
			"duplicate_bytes": es.dupBytes,
		}
	}

	result := map[string]interface{}{
		"summary": map[string]interface{}{
			"total_files":          len(allFiles),
			"total_size_bytes":     totalSize,
			"duplicate_files":      len(dupSet),
			"duplicate_size_bytes": dupSize,
			"recoverable_bytes":    dupSize,
			"scan_duration_seconds": elapsed.Seconds(),
		},
		"by_category":      byCategory,
		"by_extension":     byExtension,
		"size_distribution": sizeDist,
	}

	// Write JSON
	jsonData, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Analytics JSON error: %v\n", err)
		return
	}
	if err := os.WriteFile("analytics_go.json", jsonData, 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Analytics write error: %v\n", err)
		return
	}

	// Human-readable summary
	fmt.Println("\n--- File Type Analytics ---")
	for _, cat := range sortedKeys(byCategory) {
		entry := byCategory[cat].(map[string]interface{})
		fmt.Printf("  %-12s %5d files, %d duplicates\n",
			cat,
			entry["count"].(int),
			entry["duplicate_count"].(int),
		)
	}
	fmt.Println("Analytics written to analytics_go.json")
}

func humanBytes(b int64) string {
	const (
		kiB = 1024
		miB = 1024 * kiB
		giB = 1024 * miB
	)
	switch {
	case b >= giB:
		return fmt.Sprintf("%.2f GiB", float64(b)/float64(giB))
	case b >= miB:
		return fmt.Sprintf("%.2f MiB", float64(b)/float64(miB))
	case b >= kiB:
		return fmt.Sprintf("%.2f KiB", float64(b)/float64(kiB))
	default:
		return fmt.Sprintf("%d B", b)
	}
}

func formatSize(bytes int64) string {
	const (
		KB = 1024
		MB = 1024 * KB
		GB = 1024 * MB
	)
	switch {
	case bytes >= GB:
		return fmt.Sprintf("%.1f GB", float64(bytes)/float64(GB))
	case bytes >= MB:
		return fmt.Sprintf("%.1f MB", float64(bytes)/float64(MB))
	case bytes >= KB:
		return fmt.Sprintf("%.1f KB", float64(bytes)/float64(KB))
	default:
		return fmt.Sprintf("%d B", bytes)
	}
}
func sortedKeys(m map[string]interface{}) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: find_dups <dir1> [<dir2> ...]")
		os.Exit(1)
	}
	roots := os.Args[1:]

	startTotal := time.Now()

// ----- 1. Collect regular files (no symlinks) -----
	var allFiles []FileInfo
	var scanned int
	var totalSize int64
	var nextID = 1
	
	// Progress update struct
	type progressUpdate struct {
		count int
		size  int64
	}
	
	// Start file collection progress indicator
	progressChan := make(chan progressUpdate, 1)
	stopChan := make(chan struct{})
	go func() {
		ticker := time.NewTicker(5 * time.Second)
		defer ticker.Stop()
		lastCount := 0
		lastSize := int64(0)
		for {
			select {
			case update := <-progressChan:
				lastCount = update.count
				lastSize = update.size
			case <-ticker.C:
				if lastCount > 0 {
					fmt.Printf("\rCollecting files... %d files, %s scanned...", lastCount, formatSize(lastSize))
				}
			case <-stopChan:
				return
			}
		}
	}()

	for _, root := range roots {
		err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
			if err != nil {
				return nil // ignore access errors
			}
			if d.IsDir() {
				return nil
			}
			info, err := d.Info()
			if err != nil {
				return nil
			}
			if !info.Mode().IsRegular() {
				return nil
			}
			size := info.Size()
			if size == 0 {
				return nil // skip zero-byte files
			}
			modTime := info.ModTime()
			var birthTime time.Time
			if stat, ok := info.Sys().(*syscall.Stat_t); ok {
				birthTime = time.Unix(stat.Birthtimespec.Sec, stat.Birthtimespec.Nsec)
			} else {
				birthTime = modTime
			}
			scanned++
			totalSize += size
			progressChan <- progressUpdate{count: scanned, size: totalSize}
			allFiles = append(allFiles, FileInfo{
				ID:      nextID,
				Path:    path,
				Size:    size,
				ModTime: modTime,
				Birth:   birthTime,
			})
			nextID++
			return nil
		})
		if err != nil {
			fmt.Fprintf(os.Stderr, "Walk error %s: %v\n", root, err)
		}
	}
	
	close(stopChan)
	fmt.Printf("\rCollecting files... found %d files, %s\n", scanned, formatSize(totalSize))
	if scanned == 0 {
		return
	}

	// ----- 2. Group by size -----
	bySize := make(map[int64][]*FileInfo)
	for i := range allFiles {
		bySize[allFiles[i].Size] = append(bySize[allFiles[i].Size], &allFiles[i])
	}

	// ----- 3. Parallel hashing -----
	var filesToHash []*FileInfo
	for _, files := range bySize {
		if len(files) > 1 {
			filesToHash = append(filesToHash, files...)
		}
	}
	totalToHash := len(filesToHash)
	if totalToHash > 0 {
		fmt.Printf("Parallel hashing %d files (workers: %d)...\n", totalToHash, runtime.NumCPU())
		startHash := time.Now()

		// Start progress tracker
		progress := NewProgressTracker(totalToHash, "files")
		progress.Start()

		jobs := make(chan *FileInfo, totalToHash)
		var wg sync.WaitGroup
		for w := 0; w < runtime.NumCPU(); w++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				for f := range jobs {
					hash, err := computeSHA256(f.Path)
					if err != nil {
						fmt.Fprintf(os.Stderr, "Hash error %s: %v\n", f.Path, err)
						continue
					}
					f.Hash = hash
					progress.Increment()
				}
			}()
		}
		for _, f := range filesToHash {
			jobs <- f
		}
		close(jobs)
		wg.Wait()
		progress.Stop()
		fmt.Printf("Hashing completed in %v\n", time.Since(startHash))
	}

	// ----- 4. Build duplicate groups -----
	var csvRows [][]string
	bashLines := []string{"#!/bin/bash", "# Generated by find_dups", "set -e", ""}
	duplicatesCount := 0

	for size, files := range bySize {
		if len(files) < 2 {
			continue
		}
		hashGroups := make(map[string][]*FileInfo)
		for _, f := range files {
			if f.Hash != "" {
				hashGroups[f.Hash] = append(hashGroups[f.Hash], f)
			}
		}
		for hash, group := range hashGroups {
			if len(group) < 2 {
				continue
			}
			sort.Slice(group, func(i, j int) bool { return group[i].ID < group[j].ID })
			for _, f := range group {
				csvRows = append(csvRows, []string{
					fmt.Sprint(f.ID),
					f.Path,
					fmt.Sprint(size),
					hash,
					f.Birth.Format(time.RFC3339),
					f.ModTime.Format(time.RFC3339),
				})
			}
			for i, f := range group {
				if i == 0 {
					continue
				}
				duplicatesCount++
				bashLines = append(bashLines, fmt.Sprintf("rm -- %q", f.Path))
			}
		}
	}
	bashLines = append(bashLines, "", `echo "Deletion complete."`)

	// Write CSV
	csvFile, _ := os.Create("duplicates_go.csv")
	defer csvFile.Close()
	csvWriter := csv.NewWriter(csvFile)
	defer csvWriter.Flush()
	csvWriter.Write([]string{"FileID", "Path", "Size", "Hash", "CreationTime", "ModificationTime"})
	csvWriter.WriteAll(csvRows)

	// Write bash script
	shFile, _ := os.Create("duprm_go.sh")
	defer shFile.Close()
	for _, line := range bashLines {
		fmt.Fprintln(shFile, line)
	}
	shFile.Chmod(0755)

	// ----- 5. Sorted CSV (all files) -----
	sorted := make([]FileInfo, len(allFiles))
	copy(sorted, allFiles)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Size > sorted[j].Size })
	sortFile, _ := os.Create("sort_dup_go.csv")
	defer sortFile.Close()
	sortWriter := csv.NewWriter(sortFile)
	defer sortWriter.Flush()
	sortWriter.Write([]string{"FileID", "Path", "Size", "Hash", "CreationTime", "ModificationTime"})
	for _, f := range sorted {
		sortWriter.Write([]string{
			fmt.Sprint(f.ID),
			f.Path,
			fmt.Sprint(f.Size),
			f.Hash,
			f.Birth.Format(time.RFC3339),
			f.ModTime.Format(time.RFC3339),
		})
	}

	// ----- 6. Analytics -----
	elapsed := time.Since(startTotal)
	generateAnalytics(allFiles, bySize, elapsed)

	// ----- 7. Statistics -----
	fmt.Printf("\nFiles scanned: %d\n", scanned)
	fmt.Printf("Duplicates found (files to delete): %d\n", duplicatesCount)
	fmt.Printf("Runtime:\n")
	fmt.Printf("  - %.3f seconds\n", elapsed.Seconds())
	fmt.Printf("  - %s\n", formatDuration(elapsed))
	fmt.Println("Reports: duplicates_go.csv, sort_dup_go.csv, analytics_go.json")
	fmt.Println("Delete script: duprm_go.sh")
}

func computeSHA256(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()
	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return fmt.Sprintf("%x", h.Sum(nil)), nil
}

func formatDuration(d time.Duration) string {
	ms := d.Milliseconds()
	s := ms / 1000
	ms %= 1000
	m := s / 60
	s %= 60
	h := m / 60
	m %= 60
	return fmt.Sprintf("%02d:%02d:%02d.%03d", h, m, s, ms)
}

File: find_dups_js/find_dups.js
#!/usr/bin/env node

const fs = require('fs').promises;
const fsSync = require('fs');
const path = require('path');
const crypto = require('crypto');
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');
const os = require('os');

const NUM_WORKERS = os.cpus().length;
const BUFFER_SIZE = 65536;

const EXTENSION_CATEGORIES = {
    '.c': 'source', '.h': 'source', '.cpp': 'source', '.hpp': 'source', '.cc': 'source',
    '.cxx': 'source', '.m': 'source', '.mm': 'source', '.s': 'source', '.S': 'source',
    '.java': 'source', '.kt': 'source', '.py': 'source', '.js': 'source', '.ts': 'source',
    '.rs': 'source', '.go': 'source', '.rb': 'source', '.swift': 'source', '.sh': 'source',
    '.hex': 'firmware', '.bin': 'firmware', '.elf': 'firmware', '.dfu': 'firmware', '.flash': 'firmware', '.map': 'firmware',
    '.uvprojx': 'ide', '.uvoptx': 'ide', '.ewp': 'ide', '.eww': 'ide', '.ewt': 'ide',
    '.cproject': 'ide', '.project': 'ide', '.mxproject': 'ide', '.ioc': 'ide',
    '.yaml': 'config', '.yml': 'config', '.cmake': 'config', '.json': 'config', '.xml': 'config',
    '.conf': 'config', '.cfg': 'config', '.ini': 'config', '.toml': 'config', '.properties': 'config',
    '.pdf': 'docs', '.md': 'docs', '.txt': 'docs', '.html': 'docs', '.htm': 'docs',
    '.rst': 'docs', '.chm': 'docs', '.doc': 'docs', '.docx': 'docs', '.rtf': 'docs',
    '.png': 'image', '.jpg': 'image', '.jpeg': 'image', '.gif': 'image', '.webp': 'image',
    '.svg': 'image', '.bmp': 'image', '.ico': 'image', '.tiff': 'image', '.tif': 'image',
    '.exe': 'binary', '.dll': 'binary', '.so': 'binary', '.dylib': 'binary', '.o': 'binary', '.a': 'binary',
    '.lib': 'binary', '.obj': 'binary', '.gch': 'binary', '.pch': 'binary',
    '.zip': 'archive', '.7z': 'archive', '.tar': 'archive', '.gz': 'archive', '.bz2': 'archive',
    '.xz': 'archive', '.rar': 'archive', '.tgz': 'archive',
    '.mp4': 'media', '.wav': 'media', '.avi': 'media', '.mp3': 'media', '.ogg': 'media',
    '.flac': 'media', '.mov': 'media', '.wmv': 'media',
    '.ttf': 'font', '.otf': 'font', '.woff': 'font', '.woff2': 'font',
    '.csv': 'data', '.dts': 'data', '.dtsi': 'data', '.overlay': 'data', '.ld': 'data', '.icf': 'data', '.srec': 'data'
};

function getCategory(ext) {
    return EXTENSION_CATEGORIES[ext] || 'other';
}

async function generateAnalytics(allFiles, bySize, elapsed) {
    const catMap = {};
    const extMap = {};
    let totalSize = 0, dupFiles = 0, dupBytes = 0;
    const sizeBins = { '0_bytes': 0, under_1kb: 0, '1kb_100kb': 0, '100kb_1mb': 0, '1mb_100mb': 0, over_100mb: 0 };

    for (const f of allFiles) {
        const ext = path.extname(f.path).toLowerCase();
        const cat = getCategory(ext);
        if (!catMap[cat]) catMap[cat] = { count: 0, total_bytes: 0, duplicate_count: 0, duplicate_bytes: 0, extensions: {} };
        catMap[cat].count++;
        catMap[cat].total_bytes += f.size;
        catMap[cat].extensions[ext] = (catMap[cat].extensions[ext] || 0) + 1;
        if (!extMap[ext]) extMap[ext] = { count: 0, total_bytes: 0, duplicate_count: 0, duplicate_bytes: 0 };
        extMap[ext].count++;
        extMap[ext].total_bytes += f.size;
        totalSize += f.size;
        if (f.size === 0) sizeBins['0_bytes']++;
        else if (f.size < 1024) sizeBins.under_1kb++;
        else if (f.size < 102400) sizeBins['1kb_100kb']++;
        else if (f.size < 1048576) sizeBins['100kb_1mb']++;
        else if (f.size < 104857600) sizeBins['1mb_100mb']++;
        else sizeBins.over_100mb++;
    }

    // Identify duplicates
    for (const [size, group] of bySize.entries()) {
        if (group.length < 2) continue;
        const hashGroups = new Map();
        for (const f of group) {
            if (f.hash) {
                if (!hashGroups.has(f.hash)) hashGroups.set(f.hash, []);
                hashGroups.get(f.hash).push(f);
            }
        }
        for (const [hash, same] of hashGroups.entries()) {
            if (same.length < 2) continue;
            for (let i = 1; i < same.length; i++) {
                const f = same[i];
                const ext = path.extname(f.path).toLowerCase();
                const cat = getCategory(ext);
                catMap[cat].duplicate_count++;
                catMap[cat].duplicate_bytes += f.size;
                extMap[ext].duplicate_count++;
                extMap[ext].duplicate_bytes += f.size;
                dupFiles++;
                dupBytes += f.size;
            }
        }
    }

    const analytics = {
        summary: {
            total_files: allFiles.length,
            total_size_bytes: totalSize,
            duplicate_files: dupFiles,
            duplicate_size_bytes: dupBytes,
            recoverable_bytes: dupBytes,
            scan_duration_seconds: parseFloat(elapsed.toFixed(3))
        },
        by_category: catMap,
        by_extension: extMap,
        size_distribution: sizeBins
    };

    await fs.writeFile('analytics_js.json', JSON.stringify(analytics, null, 2));

    console.log('\n--- File Type Analytics ---');
    for (const [cat, stats] of Object.entries(catMap)) {
        console.log(`  ${cat.padEnd(10)}: ${stats.count} files, ${stats.duplicate_count} duplicates`);
    }
    console.log('Analytics written to analytics_js.json');
}

// ---------- Utility functions ----------
function csvEscape(s) {
    const str = String(s);
    if (str.includes(',') || str.includes('"') || str.includes('\n')) {
        return '"' + str.replace(/"/g, '""') + '"';
    }
    return str;
}

function formatTime(ts) {
    const d = new Date(ts);
    return d.toISOString().replace(/\.\d+Z$/, 'Z');
}

function computeSha256(filePath) {
    return new Promise((resolve, reject) => {
        const hash = crypto.createHash('sha256');
        const stream = fsSync.createReadStream(filePath, { highWaterMark: BUFFER_SIZE });
        stream.on('data', chunk => hash.update(chunk));
        stream.on('end', () => resolve(hash.digest('hex')));
        stream.on('error', reject);
    });
}

function formatDuration(seconds) {
    const msTotal = Math.floor(seconds * 1000);
    const ms = msTotal % 1000;
    let s = Math.floor(msTotal / 1000);
    const m = Math.floor(s / 60);
    s %= 60;
    const h = Math.floor(m / 60);
    const mm = m % 60;
    return `${h.toString().padStart(2,'0')}:${mm.toString().padStart(2,'0')}:${s.toString().padStart(2,'0')}.${ms.toString().padStart(3,'0')}`;
}

// Format bytes into human-readable size
function formatSize(bytes) {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let size = bytes;
    let unitIndex = 0;
    while (size >= 1024 && unitIndex < units.length - 1) {
        size /= 1024;
        unitIndex++;
    }
    return `${size.toFixed(1)} ${units[unitIndex]}`;
}

// Recursive collection of regular files (no symlinks)
async function collectFiles(roots) {
    const allFiles = [];
    let nextId = 1;
    let scanned = 0;
    let totalSize = 0;

    
    // Start progress indicator
    const progressInterval = setInterval(() => {
        if (scanned > 0) {
            process.stdout.write(`\rCollecting files... ${scanned} files, ${formatSize(totalSize)} scanned...`);
        }
    }, 5000);

    async function walk(dir) {
        let entries;
        try {
            entries = await fs.readdir(dir, { withFileTypes: true });
        } catch (err) {
            console.warn(`Warning: could not read ${dir}: ${err.message}`);
            return;
        }
        for (const entry of entries) {
            const fullPath = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                await walk(fullPath);
            } else if (entry.isFile() && !entry.isSymbolicLink()) {
                try {
                    const stat = await fs.lstat(fullPath);
                    if (!stat.isFile()) continue;
                    const size = stat.size;
                    if (size === 0) continue; // skip zero-byte files
                    totalSize += size;

                    const modTime = stat.mtimeMs;
                    const birthTime = stat.birthtimeMs;
                    allFiles.push({
                        id: nextId++,
                        path: fullPath,
                        size: size,
                        modTime: modTime,
                        birthTime: birthTime,
                        hash: null
                    });
                    scanned++;
                } catch (err) {
                    // access error, skip
                }
            }
        }
    }

    for (const root of roots) {
        const absoluteRoot = path.resolve(root);
        try {
            await fs.access(absoluteRoot);
            await walk(absoluteRoot);
        } catch (err) {
            console.warn(`Warning: directory ${absoluteRoot} inaccessible: ${err.message}`);
        }
    }
    
    process.stdout.write(`\rCollecting files... found ${scanned} files, ${formatSize(totalSize)}\n`);
    clearInterval(progressInterval);
    return { allFiles, scanned, totalSize };
}


// Hashing progress display with animated spinner
class HashingProgress {
    constructor(totalFiles) {
        this.totalFiles = totalFiles;
        this.processedFiles = 0;
        this.startTime = Date.now();
        this.intervalId = null;
        this.spinnerIndex = 0;
        this.spinChars = ['|', '/', '-', '\\'];
    }

    start() {
        this.intervalId = setInterval(() => {
            const spinner = this.spinChars[this.spinnerIndex % this.spinChars.length];
            this.spinnerIndex++;
            process.stdout.write(`\rHashing: ${this.processedFiles}/${this.totalFiles} files ${spinner}`);
        }, 1000);
    }

    increment() {
        this.processedFiles++;
    }

    stop() {
        if (this.intervalId) {
            clearInterval(this.intervalId);
            this.intervalId = null;
        }
        const elapsed = Math.floor((Date.now() - this.startTime) / 1000);
        process.stdout.write(`\rHashing: ${this.totalFiles}/${this.totalFiles} files Completed in ${elapsed}s\n`);
    }
}



// Parallel hashing via worker_threads
async function parallelHash(filesToHash) {
    const total = filesToHash.length;
    if (total === 0) return;

    console.log(`Parallel hashing ${total} files (workers: ${NUM_WORKERS})...`);
    const startHash = Date.now();

    const chunkSize = Math.ceil(total / NUM_WORKERS);
    const chunks = [];
    for (let i = 0; i < total; i += chunkSize) {
        chunks.push(filesToHash.slice(i, i + chunkSize));
    }

    // Start file-based progress tracker
    const progress = new HashingProgress(total);
    progress.start();

    // Track worker completion and accumulate file progress
    const results = await Promise.all(chunks.map(chunk => {
        return new Promise((resolve, reject) => {
            const worker = new Worker(__filename, {
                workerData: { chunk, type: 'hash' }
            });
            worker.on('message', (msg) => {
                if (msg.type === 'progress') {
                    // Incremental progress update after each file
                    progress.increment();
                } else if (msg.type === 'result') {
                    // Final result with all hashed files
                    resolve(msg.data);
                }
            });
            worker.on('error', reject);
            worker.on('exit', (code) => {
                if (code !== 0) reject(new Error(`Worker stopped with exit code ${code}`));
            });
        });
    }));

    progress.stop();

    const hashMap = new Map();
    for (const resultChunk of results) {
        for (const f of resultChunk) {
            hashMap.set(f.id, f.hash);
        }
    }
    for (const f of filesToHash) {
        const h = hashMap.get(f.id);
        if (h) f.hash = h;
    }
    console.log(`Hashing completed in ${((Date.now() - startHash) / 1000).toFixed(3)} seconds`);
}


// Worker thread for hashing
if (!isMainThread) {
    const { chunk, type } = workerData;
    if (type === 'hash') {
        (async () => {

            for (const f of chunk) {
                f.hash = await computeSha256(f.path);
                parentPort.postMessage({ type: 'progress', count: 1 });
            }
            parentPort.postMessage({ type: 'result', data: chunk });

        })().catch(err => {
            console.error(err);
            process.exit(1);
        });
    }
    return;
}

// ---------- Main function ----------
async function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.error('Usage: node find_dups.js <dir1> [<dir2> ...]');
        process.exit(1);
    }

    const startTotal = Date.now();

    // 1. Collect files
    process.stdout.write('Collecting files... ');
    const { allFiles, scanned } = await collectFiles(args);
    console.log(`found ${scanned} files`);
    if (scanned === 0) return;

    // 2. Group by size
    const bySize = new Map();
    for (const f of allFiles) {
        if (!bySize.has(f.size)) bySize.set(f.size, []);
        bySize.get(f.size).push(f);
    }

    // 3. Files to hash (size not unique)
    const filesToHash = [];
    for (const group of bySize.values()) {
        if (group.length > 1) filesToHash.push(...group);
    }

    if (filesToHash.length) {
        await parallelHash(filesToHash);
    }

    // 5. Build duplicate groups
    const csvRows = [];
    const bashLines = ['#!/bin/bash', '# Generated by find_dups_js', 'set -e', ''];
    let duplicatesCount = 0;

    for (const [size, group] of bySize.entries()) {
        if (group.length < 2) continue;
        const hashGroups = new Map();
        for (const f of group) {
            if (f.hash) {
                if (!hashGroups.has(f.hash)) hashGroups.set(f.hash, []);
                hashGroups.get(f.hash).push(f);
            }
        }
        for (const [hash, same] of hashGroups.entries()) {
            if (same.length < 2) continue;
            same.sort((a,b) => a.id - b.id);
            for (const f of same) {
                csvRows.push([
                    f.id,
                    csvEscape(f.path),
                    size,
                    hash,
                    formatTime(f.birthTime),
                    formatTime(f.modTime)
                ]);
            }
            for (let i = 1; i < same.length; i++) {
                duplicatesCount++;
                const escaped = same[i].path.replace(/'/g, "'\\''");
                bashLines.push(`rm -- '${escaped}'`);
            }
        }
    }
    bashLines.push('', 'echo "Deletion complete."');

    // 6. Write CSV and sh
    const csvHeader = ['FileID', 'Path', 'Size', 'Hash', 'CreationTime', 'ModificationTime'];
    const csvContent = [csvHeader, ...csvRows].map(row => row.join(',')).join('\n');
    await fs.writeFile('duplicates_js.csv', csvContent);
    await fs.writeFile('duprm_js.sh', bashLines.join('\n'));
    await fs.chmod('duprm_js.sh', 0o755);

    // 7. Sorted CSV of all files
    const sorted = [...allFiles].sort((a,b) => b.size - a.size);
    const sortRows = sorted.map(f => [
        f.id, csvEscape(f.path), f.size, f.hash || '',
        formatTime(f.birthTime), formatTime(f.modTime)
    ]);
    const sortContent = [csvHeader, ...sortRows].map(row => row.join(',')).join('\n');
    await fs.writeFile('sort_dup_js.csv', sortContent);

    // 8. Analytics
    const elapsed = (Date.now() - startTotal) / 1000;
    await generateAnalytics(allFiles, bySize, elapsed);

    // 9. Statistics
    console.log(`\nFiles scanned: ${scanned}`);
    console.log(`Duplicates found (files to delete): ${duplicatesCount}`);
    console.log(`Runtime:`);
    console.log(`  - ${elapsed.toFixed(3)} seconds`);
    console.log(`  - ${formatDuration(elapsed)}`);
    console.log(`Reports: duplicates_js.csv, sort_dup_js.csv, analytics_js.json`);
    console.log(`Delete script: duprm_js.sh`);
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});

